feat: add ace protocols

Will-Smith11 opened the pull request:
## 📝 Summary

Adjusted the ingestion pipeline to allow for specific interactions with known ace protocols to modify the simulation process to allow orders, in which we know are valid with a defined ace protocol through. Also some minor changes to the block building process putting protocol ace tx's that don't interact with state at the top

## 💡 Motivation and Context
Currently Other builders have started to include ace protocols and we want to also have buildernet cover this. Ace protocols have specific sequencing rules enforced on-chain that allow them to dictate execution order. By allowing for ace protocols to integrate into builders, It unlocks more transactions to execute in the same block that wouldn't be able to otherwise. The increase in transaction flow should help the builder stay competitive.
---

## Questions
This is my first time in the codebase. I wanted to open up this PR so I could get some feedback to ensure that with the given architecture that everything is implemented in the right locations. I will add a bunch of tests once I know that the architecture is solid, just don't want to do a bunch more work when I'm uncertain about design.

## ✅ I have completed the following steps:

* [x] Run `make lint`
* [x] Run `make test`
* [ ] Added tests (if applicable)

Will-Smith11 committed:
wip: ace cleanup

Will-Smith11 committed:
feat: get baseline building

Will-Smith11 committed:
feat: get ace top of block commit added

Will-Smith11 committed:
Merge develop into feat/ace-add-on

Will-Smith11 committed:
fix: sim tree

Will-Smith11 committed:
wip: almost done cleaning up sequencing

Will-Smith11 committed:
feat: finish baseline impl

Will-Smith11 committed:
fix: linting

Will-Smith11 committed:
Merge branch 'flashbots:develop' into feat/ace-add-on

Will-Smith11 committed:
chore: cleanup ace tx order type

Will-Smith11 committed:
feat: rename ace bundler

Will-Smith11 committed:
feat: wire in ace collector config

Will-Smith11 committed:
feat: just missing new detection

Will-Smith11 committed:
feat: add new detection

Will-Smith11 committed:
fix: wiring + linting

Will-Smith11 committed:
feat: more cleanup of old code

ZanCorDX committed:
removed unused error

Will-Smith11 committed:
wip: refactor to new model discussed

Will-Smith11 committed:
feat: wip

Will-Smith11 committed:
wip: some featur cleanup

Will-Smith11 committed:
feat: refactor angstrom out of the codebase, only ace

Will-Smith11 committed:
fix: comment

Will-Smith11 committed:
feat: add tests for simtree

Will-Smith11 committed:
some cleanup

Will-Smith11 committed:
cleanup

Will-Smith11 committed:
fix: slot access logic

Will-Smith11 committed:
feat: merge develop + some basic refactor

Will-Smith11 committed:
feat: refactor sim_job

Will-Smith11 committed:
feat: cleanup simjob

Will-Smith11 committed:
fix: clippy

Will-Smith11 committed:
feat: add ability to disable ace + remove hacky code

Will-Smith11 committed:
make sure that clippy happy

Will-Smith11 committed:
bug fixes

Will-Smith11 committed:
cleanup ace enabled flag

Will-Smith11 committed:
fix: remove unused code

Will-Smith11 committed:
more cleanup

Will-Smith11 committed:
feat: test and cases

Will-Smith11 committed:
cleanup from bad merge

Will-Smith11 committed:
wip

Will-Smith11 committed:
bug fixes

Will-Smith11 committed:
fix edgecases

Will-Smith11 committed:
fix: remove comment thats misleading

ZanCorDX commented on crates/rbuilder/src/building/sim.rs:
```diff
@@ -299,14 +531,163 @@ impl SimTree {
+    /// Handle ACE unlocking interaction after successful simulation.
+    /// Returns optional cancellation OrderId if a mempool unlock cancels an optional ACE tx.
+    /// Note: NonUnlocking ACE interactions are handled at the OrderSimResult level.
+    pub fn handle_ace_unlock(
```
Should take as args simulated_order,previous_orders since it's always called in match SimulatedResult::Success no need to recheck that.

ZanCorDX commented on crates/rbuilder/src/building/sim.rs:
```diff
@@ -299,14 +531,163 @@ impl SimTree {
+    pub fn handle_ace_unlock(
+        &mut self,
+        result: &SimulatedResult,
+    ) -> Result<Vec<OrderId>, ProviderError> {
```
Never returns error!!! Remove Result

ZanCorDX commented on crates/rbuilder/src/building/sim.rs:
```diff
@@ -299,14 +531,163 @@ impl SimTree {
+                            "Cancelling optional ACE unlock for {:?} - user unlock exists",
+                            contract_address
+                        );
+                        cancellations.push(simulated_order.order.id());
```
It's cancelling it's own order?
It should only cancell the order returned by mark_mempool_unlock.
If it wants to express that the order analized should not be inserted it should return something like a bool saying it so we don’t send and order and it’s cancellation.

ZanCorDX commented on crates/rbuilder/src/building/sim.rs:
```diff
@@ -299,14 +531,163 @@ impl SimTree {
+        };
+
+        // If this order already has parents, it was re-simulated - just pass through
+        if !previous_orders.is_empty() {
```
Will return for a user tx having a previous nonce dep, are we ok with this?

ZanCorDX commented on crates/rbuilder/src/building/sim.rs:
```diff
@@ -299,14 +531,163 @@ impl SimTree {
+                        contract_address
+                    );
+                }
+                AceUnlockSource::User => {
```
There is NO guaranty that a AceUnlockSource::User will be executed before orders depending o it.
A depending order could make a lot of money so it could be placed before the AceUnlockSource::User.

Will-Smith11 replied:
This should be fine. otherwise we are making a lot more downstream changes

ZanCorDX commented on crates/rbuilder/src/building/sim.rs:
```diff
@@ -299,14 +531,163 @@ impl SimTree {
         &mut self,
         results: Vec<SimulatedResult>,
-    ) -> Result<(), ProviderError> {
+    ) -> Result<(Vec<SimulatedResult>, Vec<OrderId>), ProviderError> {
```
nobody uses the Vec<SimulatedResult> in the result. Remove it.

ZanCorDX commented on crates/rbuilder/src/building/sim.rs:
```diff
@@ -369,36 +751,61 @@ where
+                    if let Some(contract_address) = failure.ace_dependency {
+                        // Order failed but needs ACE unlock - queue for re-simulation
+                        // Pass existing ace_unlock_contracts for progressive multi-ACE discovery
+                        sim_tree.add_ace_dependency_for_order(
```
Why simulate_all_orders_with_sim_tree is calling "sim_tree.add_ace_dependency_for_order" manually? Isn't submit_simulation_tasks_results doing that already?

ZanCorDX commented on crates/rbuilder/src/building/sim.rs:
```diff
@@ -276,21 +500,29 @@ impl SimTree {
+            let pending_state = self.get_order_dependency_state(&ready_pending_order.order)?;
             match pending_state {
-                OrderNonceState::Ready(parents) => {
+                OrderDependencyState::Ready(mut parents) => {
```
Here parents contains only Nonce dependencies and not any other prev ACE dependency.
If the order was depending in another ACE dependency that will not be included.
If looks like you should add all deps in PendingOrder::ace_unlock_contracts not only this one last unlocked.

ZanCorDX commented on crates/rbuilder/src/building/sim.rs:
```diff
@@ -47,28 +64,101 @@ pub struct NonceKey {
     order: Order,
-    unsatisfied_nonces: usize,
+    unsatisfied_dependencies: usize,
+    /// ACE contracts already provided as unlock parents (for progressive multi-ACE discovery)
```
Shouldn't this commet just be "ACE contract needed to be unlocked"?
You can have this pending without alread having an unlock parent.....

ZanCorDX commented on crates/rbuilder/src/building/sim.rs:
```diff
@@ -187,27 +313,96 @@ impl SimTree {
+        // Check if we already have an unlock provider for the new contract
+        if self.dependency_providers.contains_key(&dep_key) {
+            // Build parents from ALL ACE unlock contracts we need
+            let mut parents = Vec::new();
```
You are not adding Nonce deps here.

ZanCorDX commented:
From the comment I made in add_ace_dependency_for_order and process_simulation_task_result it feels that you need a common func that computes the parent.
Maybe get_order_dependency_state gets renamed to get_order_nonce_dependency_state (since it only computes nonce deps and the new function get_order_dependency_state takes an extra parameter with the ACE needed contracts and also calls get_order_dependency_state ?

ZanCorDX commented on crates/rbuilder/src/building/sim.rs:
```diff
@@ -453,32 +872,138 @@ pub fn simulate_order_using_fork<Tracer: SimulationTracer>(
+        // Use HashMap to track best interaction per contract
+        let mut per_contract: HashMap<Address, AceInteraction> = HashMap::default();
+
+        for (tx, _) in order.list_txs() {
```
All txts are being iterated, even if they where droped

ZanCorDX commented on crates/rbuilder/src/building/sim.rs:
```diff
@@ -453,32 +872,138 @@ pub fn simulate_order_using_fork<Tracer: SimulationTracer>(
+            // Check this transaction against all ACE configs
+            for (_, config) in ace_configs.iter() {
+                if let Some(interaction) =
+                    classify_ace_interaction(trace, sim_success, config, selector, tx_to, tx_from)
```
sim_success is for the whole Order, it's not related to the particular execution of this Tx.
This might be a problem since we don't have at the moment trace state per child tx.
Eg: For a Bundle we could have a Dropeable/Revertable Tx accessing the pool and failing but the bundle succeding so we would have classify_ace_interaction answering Unlocking.
To detect Force/Optional unlocking orders we could asume that Order must be a single Tx.

Will-Smith11 committed:
merge upstream

Will-Smith11 committed:
wip: refactor

Will-Smith11 committed:
cleanup

Will-Smith11 committed:
feat: update config examples

Will-Smith11 committed:
add logging

Will-Smith11 committed:
feat: proper accounting for no fees

Will-Smith11 committed:
feat: proper classification of ace user orders

Will-Smith11 committed:
test, inject on-chain landed orders

Will-Smith11 committed:
some better classification

Will-Smith11 committed:
Merge branch 'develop' into feat/ace-add-on

; Conflicts:
;	crates/rbuilder/src/backtest/execute.rs
;	crates/rbuilder/src/bin/run-bundle-on-prefix.rs
;	crates/rbuilder/src/building/block_orders/order_priority.rs
;	crates/rbuilder/src/building/block_orders/test_data_generator.rs
;	crates/rbuilder/src/building/builders/parallel_builder/block_building_result_assembler.rs
;	crates/rbuilder/src/building/builders/parallel_builder/conflict_resolvers.rs
;	crates/rbuilder/src/building/builders/parallel_builder/conflict_task_generator.rs
;	crates/rbuilder/src/building/builders/parallel_builder/groups.rs
;	crates/rbuilder/src/building/sim.rs
;	crates/rbuilder/src/building/testing/bundle_tests/setup.rs
;	crates/rbuilder/src/live_builder/base_config.rs
;	crates/rbuilder/src/live_builder/building/mod.rs
;	crates/rbuilder/src/live_builder/mod.rs

Will-Smith11 committed:
fix: use ACE priority instead of profit when comparing unlock dependency providers

The Entry::Occupied branch in SimTree::process_simulation_task_result compared
dependency providers by coinbase_profit for all dep types, allowing a more
profitable optional unlock to displace a force unlock. Now AceUnlock deps
compare by interaction priority (Force > Optional > User) while Nonce deps
retain profit-based comparison.

Co-Authored-By: Frankenstein <263438330+frankenstien-dev@users.noreply.github.com>

Will-Smith11 committed:
Merge remote-tracking branch 'upstream/develop' into feat/ace-add-on

Will-Smith11 committed:
w

Will-Smith11 committed:
fix: drain parent traces before ACE classification in fork simulation

When simulating an order with ACE unlock parents, the tracer accumulated
state traces from both parent and child execution. This caused the child
order to be misclassified as Unlocking { source: User } when it merely
reads the unlocked detection slot, because parent writes leaked into the
combined trace used for classification.

Adds a has_ace_parents flag to SimulationRequest that, when set, drains
the accumulated parent traces before executing the target order so ACE
interaction classification reflects only the order's own slot access.

Will-Smith11 committed:
Merge branch 'develop' into feat/ace-add-on

; Conflicts:
;	crates/rbuilder/src/building/sim.rs
;	crates/rbuilder/src/live_builder/simulation/sim_worker.rs
;	crates/rbuilder/src/live_builder/simulation/simulation_job.rs

Will-Smith11 committed:
fix: preserve develop's single-nonce-only provider tracking in SimTree

Only register nonce dependencies as providers when a simulation updates
exactly one nonce, matching develop's sims_that_update_one_nonce behavior.
Multi-nonce sims (e.g. bundles with multiple senders) are stored but not
tracked as providers. ACE unlock deps are always processed regardless.
